src/net_processing.cpp

```diff
@@ -584,76 +578,105 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
+            xver.set_u64c(XVer::BU_GRAPHENE_MIN_VERSION_SUPPORTED, grapheneMinVersionSupported.Value());
+            xver.set_u64c(XVer::BU_GRAPHENE_FAST_FILTER_PREF, grapheneFastFilterCompatibility.Value());
+            xver.set_u64c(XVer::BU_MEMPOOL_SYNC, syncMempoolWithPeers.Value());
+            xver.set_u64c(XVer::BU_MEMPOOL_SYNC_MAX_VERSION_SUPPORTED, mempoolSyncMaxVersionSupported.Value());
+            xver.set_u64c(XVer::BU_MEMPOOL_SYNC_MIN_VERSION_SUPPORTED, mempoolSyncMinVersionSupported.Value());
+            xver.set_u64c(XVer::BU_XTHIN_VERSION, 2); // xthin version
 
-        xver.set_u64c(XVer::BU_MEMPOOL_ANCESTOR_COUNT_LIMIT, nLimitAncestors);
-        xver.set_u64c(XVer::BU_MEMPOOL_ANCESTOR_SIZE_LIMIT, nLimitAncestorSize);
-        xver.set_u64c(XVer::BU_MEMPOOL_DESCENDANT_COUNT_LIMIT, nLimitDescendants);
-        xver.set_u64c(XVer::BU_MEMPOOL_DESCENDANT_SIZE_LIMIT, nLimitDescendantSize);
-        xver.set_u64c(XVer::BU_TXN_CONCATENATION, 1);
+            size_t nLimitAncestors = GetArg("-limitancestorcount", BU_DEFAULT_ANCESTOR_LIMIT);
+            size_t nLimitAncestorSize = GetArg("-limitancestorsize", BU_DEFAULT_ANCESTOR_SIZE_LIMIT) * 1000;
+            size_t nLimitDescendants = GetArg("-limitdescendantcount", BU_DEFAULT_DESCENDANT_LIMIT);
+            size_t nLimitDescendantSize = GetArg("-limitdescendantsize", BU_DEFAULT_DESCENDANT_SIZE_LIMIT) * 1000;
 
-        electrum::set_xversion_flags(xver, chainparams.NetworkIDString());
+            xver.set_u64c(XVer::BU_MEMPOOL_ANCESTOR_COUNT_LIMIT, nLimitAncestors);
+            xver.set_u64c(XVer::BU_MEMPOOL_ANCESTOR_SIZE_LIMIT, nLimitAncestorSize);
+            xver.set_u64c(XVer::BU_MEMPOOL_DESCENDANT_COUNT_LIMIT, nLimitDescendants);
+            xver.set_u64c(XVer::BU_MEMPOOL_DESCENDANT_SIZE_LIMIT, nLimitDescendantSize);
+            xver.set_u64c(XVer::BU_TXN_CONCATENATION, 1);
 
-        pfrom->PushMessage(NetMsgType::XVERSION, xver);
+            electrum::set_xversion_flags(xver, chainparams.NetworkIDString());
 
+            pfrom->PushMessage(NetMsgType::XVERSION_OLD, xver);
+        }
 
-        if (pfrom->nVersion >= EXPEDITED_VERSION)
-            // also send legacy message (at least for now)
-            pfrom->PushMessage(NetMsgType::BUVERSION, GetListenPort());
-        pfrom->state_outgoing = ConnectionStateOutgoing::READY;
+        handleAddressAfterInit(pfrom);
+        enableSendHeaders(pfrom);
+        enableCompactBlocks(pfrom);
 
         // Tell the peer what maximum xthin bloom filter size we will consider acceptable.
         // FIXME: integrate into xversion as well
         if (pfrom->ThinBlockCapable() && IsThinBlocksEnabled())
         {
             pfrom->PushMessage(NetMsgType::FILTERSIZEXTHIN, nXthinBloomFilterSize);
         }
-    }
 
+        // This step done after final handshake
+        CheckAndRequestExpeditedBlocks(pfrom);
+
+        pfrom->fSuccessfullyConnected = true;
```

Comment:
shouldn't we move fSuccesfullyConnected=true to the beginning of this scope...so as soon as we receive the VERACK then we should be connected...then send out all the other notifications which may result in data begin returned?  Otherwise it could be possible to receive messages before setting the flag.